File: src/cyrillic.ts
import { firstCase, isUpper, lowerChar, upperText } from './case.js';
import seedForeignWords from './data/foreign-words.json';
import { oldLatinCore, toNewLatinMapped } from './latin.js';
import { normalizeConfusables, prepareTextMapped, sourceRange } from './normalize.js';
import { applyExceptions, mapSegments, splitProtected } from './pipeline.js';
import type { ConversionOptions, ConversionResult, Warning } from './types.js';

const DIRECT_FROM_CYRILLIC: Readonly<Record<string, string>> = {
  а: 'a',
  б: 'b',
  в: 'v',
  г: 'g',
  д: 'd',
  ж: 'j',
  з: 'z',
  и: 'i',
  й: 'y',
  к: 'k',
  л: 'l',
  м: 'm',
  н: 'n',
  о: 'o',
  п: 'p',
  р: 'r',
  с: 's',
  т: 't',
  у: 'u',
  ф: 'f',
  х: 'x',
  ч: 'ç',
  ш: 'ş',
  ы: 'i',
  э: 'e',
  ў: 'ö',
  қ: 'q',
  ғ: 'ğ',
  ҳ: 'h',
};

const DIRECT_TO_CYRILLIC: Readonly<Record<string, string>> = {
  a: 'а',
  b: 'б',
  v: 'в',
  g: 'г',
  d: 'д',
  j: 'ж',
  z: 'з',
  i: 'и',
  y: 'й',
  k: 'к',
  l: 'л',
  m: 'м',
  n: 'н',
  o: 'о',
  p: 'п',
  r: 'р',
  s: 'с',
  t: 'т',
  u: 'у',
  f: 'ф',
  x: 'х',
  ç: 'ч',
  ş: 'ш',
  ö: 'ў',
  q: 'қ',
  ğ: 'ғ',
  h: 'ҳ',
};

const CYRILLIC_VOWELS = new Set(['а', 'е', 'ё', 'и', 'о', 'у', 'ў', 'э', 'ю', 'я']);

const LATIN_VOWELS = new Set(['a', 'e', 'i', 'o', 'u', 'ö']);
// Vowels after which ь becomes a written y; iotated vowels (ё ю я е) already supply it.
const SOFT_SIGN_GLIDE_VOWELS = new Set(['а', 'о', 'у', 'э', 'ў']);

function warning(
  index: number,
  rule: string,
  message: string,
  alternatives: string[],
  length = 1,
): Warning {
  return { index, length, rule, message, alternatives };
}

/** Rewrite warning ranges from prepared-text offsets to offsets in the caller's text. */
function toSourceOffsets(
  warnings: Warning[],
  origins: readonly number[] | null,
  offset: number,
): void {
  for (const item of warnings) {
    const range = sourceRange(origins, item.index, item.length);
    item.index = offset + range.index;
    item.length = range.length;
  }
}

function isLatinLetter(character: string): boolean {
  return (
    /^\p{L}$/u.test(character) &&
    /^[A-Za-z\u00c0-\u024f\u1e00-\u1eff]/u.test(character.normalize('NFKC'))
  );
}

/** The lowercased letter at `index` for positional rules, or '' when it is not a letter. */
function letterAt(source: string, index: number): string {
  if (index < 0 || index >= source.length) return '';
  const unit = source.charCodeAt(index);
  const before = index > 0 ? source.charCodeAt(index - 1) : 0;
  if (unit >= 0xdc00 && unit <= 0xdfff && before >= 0xd800 && before <= 0xdbff) {
    return letterAt(source, index - 1);
  }
  const codePoint = String.fromCodePoint(source.codePointAt(index)!);
  return /\p{L}/u.test(codePoint) ? lowerChar(codePoint) : '';
}

/** Case a multi-letter replacement: all caps inside an uppercase word, else title case. */
function caseReplacement(source: string, index: number, lower: string): string {
  const character = source[index]!;
  if (!isUpper(character)) return lower;
  const inUppercaseWord =
    isUpper(source[index + 1] ?? '') || (index > 0 && isUpper(source[index - 1]!));
  return inUppercaseWord ? upperText(lower) : firstCase(character, lower);
}

function fromCyrillicCore(
  text: string,
  offset: number,
  options: ConversionOptions,
): ConversionResult {
  const { text: source, origins } = applyExceptions(prepareTextMapped(text), options);
  let output = '';
  const warnings: Warning[] = [];
  let previousLetter = '';

  for (let index = 0; index < source.length; index += 1) {
    const character = source[index]!;
    const lower = lowerChar(character);
    let replacement: string | undefined;

    if (lower === 'е') {
      const useYe =
        previousLetter === '' ||
        CYRILLIC_VOWELS.has(previousLetter) ||
        previousLetter === 'ъ' ||
        previousLetter === 'ь';
      replacement = useYe ? 'ye' : 'e';
      warnings.push(
        warning(
          index,
          'cyrillic.e.positional',
          'Cyrillic е can represent e or ye; a positional rule was applied.',
          ['e', 'ye'],
        ),
      );
    } else if (lower === 'ц') {
      const useS = previousLetter === '' || previousLetter === 'ц';
      replacement = useS ? 's' : 'ts';
      warnings.push(
        warning(
          index,
          'cyrillic.tse.positional',
          'Cyrillic ц can represent s or ts; a positional rule was applied.',
          ['s', 'ts'],
        ),
      );
    } else if (lower === 'щ') {
      replacement = 'şç';
      warnings.push(
        warning(
          index,
          'cyrillic.shcha.ambiguous',
          'Cyrillic щ has no single new-Latin equivalent; şç was chosen.',
          ['şç', 'ş'],
        ),
      );
    } else if (lower === 'ъ') {
      replacement = 'ʼ';
      warnings.push(
        warning(
          index,
          'cyrillic.hard-sign.ambiguous',
          'The hard sign was retained as tutuq belgisi.',
          ['ʼ', ''],
        ),
      );
    } else if (lower === 'ь') {
      // Before a plain vowel the soft sign marks a glide (батальон → batalyon); elsewhere it
      // has no Latin counterpart and is dropped (медаль → medal).
      const glide = SOFT_SIGN_GLIDE_VOWELS.has(letterAt(source, index + 1));
      replacement = glide ? 'y' : '';
      warnings.push(
        warning(
          index,
          'cyrillic.soft-sign.ambiguous',
          glide ? 'The soft sign before a vowel was written as y.' : 'The soft sign was dropped.',
          glide ? ['y', ''] : ['', 'ʼ'],
        ),
      );
    } else if (lower === 'ё' || lower === 'ю' || lower === 'я') {
      replacement = lower === 'ё' ? 'yo' : lower === 'ю' ? 'yu' : 'ya';
      const name = lower === 'ё' ? 'yo' : lower === 'ю' ? 'yu' : 'ya';
      warnings.push(
        warning(index, `cyrillic.${name}.compound`, `Cyrillic ${lower} was expanded to ${name}.`, [
          name,
          name.slice(1),
        ]),
      );
    } else {
      replacement = DIRECT_FROM_CYRILLIC[lower];
    }

    output += replacement === undefined ? character : caseReplacement(source, index, replacement);
    previousLetter = letterAt(source, index);
  }
  toSourceOffsets(warnings, origins, offset);
  return { text: oldLatinCore(output, options).normalize('NFC'), warnings };
}

/** Convert Uzbek Cyrillic to new Latin and report every lossy or ambiguous choice. */
export function fromCyrillic(text: string, options: ConversionOptions = {}): ConversionResult {
  return mapSegments(text, options, (segment, start) => fromCyrillicCore(segment, start, options));
}

/** Lowercase letters of the 2026 Latin alphabet. Any other Latin letter marks a foreign word. */
const UZBEK_LATIN_LETTERS = new Set([...'abdefghijklmnopqrstuvxyzöğşç']);

/**
 * Brand and product names whose letters are all Uzbek letters, so nothing in the spelling gives
 * them away, yet whose Latin form is the norm in Uzbek text (`google` letter by letter is
 * `гоогле`). Matched as whole words only, optionally with an Uzbek case or plural suffix.
 */
const KNOWN_FOREIGN_NAMES = new Set<string>(seedForeignWords);
const UZBEK_SUFFIXES = [
  'da',
  'ga',
  'ni',
  'ning',
  'dan',
  'lar',
  'larda',
  'larga',
  'larni',
  'larning',
  'lardan',
  'dagi',
  'lardagi',
  'gacha',
  'mi',
];

// What continues a word in the foreign-word scan: letters, marks, digits, hyphens and every
// apostrophe-like character, so `Wi-Fi`, `oʻzbek`, `o'zbek` and `McDonald's` are single words.
const WORD_PART = /^[\p{L}\p{M}\p{N}ʻʼ'‘’`´′＇-]$/u;
const EDGE_PUNCTUATION = /^[ʻʼ'‘’`´′＇-]+|[ʻʼ'‘’`´′＇-]+$/gu;

function isKnownForeignName(word: string): boolean {
  const lower = [...word.replace(EDGE_PUNCTUATION, '')].map(lowerChar).join('');
  // `google-da` counts through its first part; `googleda` through a suffix.
  for (const candidate of new Set([lower, lower.split('-')[0]!])) {
    if (KNOWN_FOREIGN_NAMES.has(candidate)) return true;
    for (const suffix of UZBEK_SUFFIXES) {
      if (
        candidate.length > suffix.length &&
        candidate.endsWith(suffix) &&
        KNOWN_FOREIGN_NAMES.has(candidate.slice(0, -suffix.length))
      )
        return true;
    }
  }
  return false;
}

type ForeignReason = 'letters' | 'name' | null;

/**
 * A word is foreign when it has a Latin letter outside the Uzbek alphabet (`w`, `ü`, a `c`
 * that is not part of the previous alphabet's `ch`) or a capital after a lowercase letter
 * (`iPhone`, `SiO2`), or when it is a known brand name. Uzbek words never do either, so
 * converting such a word letter by letter would only produce a mix of two alphabets. Astral
 * letters are skipped for parity with the Python port.
 */
function foreignReason(word: string): ForeignReason {
  const characters = [...normalizeConfusables(word)];
  let afterLowercase = false;
  for (let at = 0; at < characters.length; at += 1) {
    const character = characters[at]!;
    if (character.length > 1 || !isLatinLetter(character)) {
      afterLowercase = false;
      continue;
    }
    const lower = lowerChar(character);
    const digraph = lower === 'c' && lowerChar(characters[at + 1] ?? '') === 'h';
    if (!digraph && !UZBEK_LATIN_LETTERS.has(lower)) return 'letters';
    if (isUpper(character) && afterLowercase) return 'letters';
    afterLowercase = character !== upperText(character);
  }
  return isKnownForeignName(word) ? 'name' : null;
}

interface Piece {
  text: string;
  start: number;
  foreign: ForeignReason;
}

/** Cut `text` (unprotected, starting at `start` in the caller's text) around foreign words. */
function splitForeignWords(text: string, start: number): Piece[] {
  const pieces: Piece[] = [];
  let cursor = 0;
  let index = 0;
  while (index < text.length) {
    const character = String.fromCodePoint(text.codePointAt(index)!);
    if (!WORD_PART.test(character)) {
      index += character.length;
      continue;
    }
    let end = index;
    while (end < text.length) {
      const part = String.fromCodePoint(text.codePointAt(end)!);
      if (!WORD_PART.test(part)) break;
      end += part.length;
    }
    const word = text.slice(index, end);
    const reason = foreignReason(word);
    if (reason !== null) {
      if (index > cursor)
        pieces.push({ text: text.slice(cursor, index), start: start + cursor, foreign: null });
      pieces.push({ text: word, start: start + index, foreign: reason });
      cursor = end;
    }
    index = end;
  }
  if (cursor < text.length)
    pieces.push({ text: text.slice(cursor), start: start + cursor, foreign: null });
  return pieces;
}

/** Convert the letters at `index` and report the choice; returns the output piece and units read. */
function convertAt(
  source: string,
  index: number,
  warnings: Warning[],
): { piece: string; consumed: number } {
  const character = source[index]!;
  const lower = lowerChar(character);
  const uppercase = character !== lower;
  const pair = lower + lowerChar(source[index + 1] ?? '');
  // Inverse of the fromCyrillic е rule: Cyrillic е is read "ye" at a word start, after
  // a vowel, or after ъ/ь, so there "ye" maps back to е and a bare "e" must be э.
  const previous = index > 0 ? letterAt(source, index - 1) : '';
  const positional = previous === '' || previous === 'ʼ' || LATIN_VOWELS.has(previous);
  let replacement: string | undefined;
  let consumed = 1;

  if (pair === 'şç') {
    replacement = 'щ';
    consumed = 2;
    warnings.push(
      warning(
        index,
        'latin.shcha.ambiguous',
        'şç was interpreted as Cyrillic щ rather than шч.',
        ['щ', 'шч'],
        2,
      ),
    );
  } else if (pair === 'yo' || pair === 'yu' || pair === 'ya') {
    replacement = pair === 'yo' ? 'ё' : pair === 'yu' ? 'ю' : 'я';
    consumed = 2;
    warnings.push(
      warning(
        index,
        `latin.${pair}.ambiguous`,
        `${pair} was interpreted as one iotated Cyrillic letter.`,
        [replacement, `й${DIRECT_TO_CYRILLIC[pair[1]!]}`],
        2,
      ),
    );
  } else if (pair === 'ts') {
    replacement = 'ц';
    consumed = 2;
    warnings.push(
      warning(index, 'latin.tse.ambiguous', 'ts was interpreted as Cyrillic ц.', ['ц', 'тс'], 2),
    );
  } else if (pair === 'ye' && positional) {
    replacement = 'е';
    consumed = 2;
    warnings.push(
      warning(
        index,
        'latin.ye.positional',
        'ye at a word start, after a vowel, or after tutuq was interpreted as Cyrillic е.',
        ['е', 'йе'],
        2,
      ),
    );
  } else if (lower === 'e') {
    replacement = positional ? 'э' : 'е';
    warnings.push(
      warning(
        index,
        'latin.e.ambiguous',
        `Latin e can correspond to Cyrillic е or э; ${replacement} was chosen.`,
        positional ? ['э', 'е'] : ['е', 'э'],
      ),
    );
  } else if (character === 'ʼ') {
    replacement = 'ъ';
    warnings.push(
      warning(index, 'latin.tutuq.ambiguous', 'Tutuq belgisi was interpreted as a hard sign.', [
        'ъ',
        'ь',
        '',
      ]),
    );
  } else if (lower === 'c') {
    replacement = 'ц';
    warnings.push(
      warning(index, 'latin.c.ambiguous', 'Standalone c was interpreted as Cyrillic ц.', [
        'ц',
        'с',
      ]),
    );
  } else {
    replacement = DIRECT_TO_CYRILLIC[lower];
    if (replacement === undefined && (isLatinLetter(character) || character === 'ʻ')) {
      warnings.push(
        warning(
          index,
          'latin.unmapped',
          character === 'ʻ'
            ? 'Stray ʻ is not part of oʻ or gʻ and has no Cyrillic mapping.'
            : `Latin ${character} has no Cyrillic mapping.`,
          [],
        ),
      );
    }
  }

  const piece =
    replacement === undefined
      ? character
      : uppercase
        ? firstCase(character, replacement)
        : replacement;
  return { piece, consumed };
}

function toCyrillicCore(text: string, offset: number): ConversionResult {
  const { text: source, origins } = prepareTextMapped(text);
  let output = '';
  const warnings: Warning[] = [];
  let index = 0;
  while (index < source.length) {
    const step = convertAt(source, index, warnings);
    output += step.piece;
    index += step.consumed;
  }
  toSourceOffsets(warnings, origins, offset);
  return { text: output.normalize('NFC'), warnings };
}

/** Latin → Cyrillic for one span with no foreign words; warning offsets are shifted by `start`. */
function latinToCyrillic(text: string, options: ConversionOptions, start = 0): ConversionResult {
  const canonical = toNewLatinMapped(text, options);
  const result = mapSegments(canonical.text, options, (segment, offset) =>
    toCyrillicCore(segment, offset),
  );
  // Warnings point into the new-Latin intermediate; report them against the caller's text.
  toSourceOffsets(result.warnings, canonical.origins, start);
  return result;
}

/**
 * Convert new (or previous) Uzbek Latin to Cyrillic and report every lossy or ambiguous choice.
 * Words the Uzbek alphabet cannot spell (`Windows`, `Microsoft`, `iPhone`) are kept as written
 * and reported as `latin.foreign` unless `foreignWords` is `'transliterate'`; so are known brand
 * names such as `google` (see data/foreign-words.json).
 */
export function toCyrillic(text: string, options: ConversionOptions = {}): ConversionResult {
  const keepForeign = options.foreignWords !== 'transliterate';
  let output = '';
  const warnings: Warning[] = [];
  for (const segment of splitProtected(text, options)) {
    if (segment.protected) {
      output += segment.text;
      continue;
    }
    const pieces = keepForeign
      ? splitForeignWords(segment.text, segment.start)
      : [{ text: segment.text, start: segment.start, foreign: null }];
    for (const piece of pieces) {
      if (piece.foreign !== null) {
        output += piece.text;
        warnings.push(
          warning(
            piece.start,
            'latin.foreign',
            piece.foreign === 'name'
              ? `${piece.text} is a known foreign name and was left unchanged.`
              : `${piece.text} looks foreign (non-Uzbek letters or mixed case) and was left unchanged.`,
            [piece.text, latinToCyrillic(piece.text, options).text],
            piece.text.length,
          ),
        );
        continue;
      }
      const converted = latinToCyrillic(piece.text, options, piece.start);
      output += converted.text;
      for (const item of converted.warnings) warnings.push(item);
    }
  }
  return { text: output, warnings };
}

File: python/src/alifbo/cyrillic.py
"""Cyrillic <-> new Latin conversion with warnings (mirrors ``src/cyrillic.ts``)."""

from __future__ import annotations

import re
import unicodedata
from typing import Iterable, List, Mapping, Optional, Sequence, Tuple

from .case import first_case, is_upper, lower_char, upper_text
from .latin import old_latin_core, to_new_latin_mapped
from .normalize import nfc, prepare_text_mapped, source_range, unit_length
from .pipeline import Options, apply_exceptions, make_options, map_segments
from .types import ConversionResult, Warning

_DIRECT_FROM_CYRILLIC = {
    "а": "a",
    "б": "b",
    "в": "v",
    "г": "g",
    "д": "d",
    "ж": "j",
    "з": "z",
    "и": "i",
    "й": "y",
    "к": "k",
    "л": "l",
    "м": "m",
    "н": "n",
    "о": "o",
    "п": "p",
    "р": "r",
    "с": "s",
    "т": "t",
    "у": "u",
    "ф": "f",
    "х": "x",
    "ч": "ç",
    "ш": "ş",
    "ы": "i",
    "э": "e",
    "ў": "ö",
    "қ": "q",
    "ғ": "ğ",
    "ҳ": "h",
}

_DIRECT_TO_CYRILLIC = {
    "a": "а",
    "b": "б",
    "v": "в",
    "g": "г",
    "d": "д",
    "j": "ж",
    "z": "з",
    "i": "и",
    "y": "й",
    "k": "к",
    "l": "л",
    "m": "м",
    "n": "н",
    "o": "о",
    "p": "п",
    "r": "р",
    "s": "с",
    "t": "т",
    "u": "у",
    "f": "ф",
    "x": "х",
    "ç": "ч",
    "ş": "ш",
    "ö": "ў",
    "q": "қ",
    "ğ": "ғ",
    "h": "ҳ",
}

_CYRILLIC_VOWELS = frozenset(("а", "е", "ё", "и", "о", "у", "ў", "э", "ю", "я"))
_LATIN_VOWELS = frozenset(("a", "e", "i", "o", "u", "ö"))
# Vowels after which ь becomes a written y; iotated vowels (ё ю я е) already supply it.
_SOFT_SIGN_GLIDE_VOWELS = frozenset(("а", "о", "у", "э", "ў"))
_IOTATED = {"ё": "yo", "ю": "yu", "я": "ya"}
_IOTATED_REVERSE = {"yo": "ё", "yu": "ю", "ya": "я"}


def _warning(
    index: int, rule: str, message: str, alternatives: Sequence[str], length: int = 1
) -> Warning:
    return Warning(index, length, rule, message, tuple(alternatives))


def _to_source_offsets(warnings: List[Warning], origins: List[int], offset: int) -> List[Warning]:
    """Rewrite warning ranges from prepared-text offsets to offsets in the caller's text."""
    mapped: List[Warning] = []
    for item in warnings:
        index, length = source_range(origins, item.index, item.length)
        mapped.append(Warning(offset + index, length, item.rule, item.message, item.alternatives))
    return mapped


def _letter_at(source: str, index: int) -> str:
    """The lowercased letter at ``index`` for positional rules, or '' when it is not a letter."""
    if index < 0 or index >= len(source):
        return ""
    character = source[index]
    return lower_char(character) if unicodedata.category(character)[0] == "L" else ""


_LATIN_START = re.compile("[A-Za-z\u00c0-\u024f\u1e00-\u1eff]")


def _is_latin_letter(character: str) -> bool:
    """True for a BMP letter whose NFKC form starts in Latin-1, Extended-A/B or Additional.

    Astral letters (e.g. U+1D400) are skipped: the TypeScript loop walks UTF-16
    code units, so a surrogate half never matches ``\\p{L}``.
    """
    if unicodedata.category(character)[0] != "L" or ord(character) > 0xFFFF:
        return False
    return _LATIN_START.match(unicodedata.normalize("NFKC", character)) is not None


def _case_replacement(source: str, index: int, lower: str) -> str:
    """Case a multi-letter replacement: all caps inside an uppercase word, else title case."""
    character = source[index]
    if not is_upper(character):
        return lower
    in_uppercase_word = (index + 1 < len(source) and is_upper(source[index + 1])) or (
        index > 0 and is_upper(source[index - 1])
    )
    return upper_text(lower) if in_uppercase_word else first_case(character, lower)


def _from_cyrillic_core(text: str, offset: int, options: Options) -> Tuple[str, List[Warning]]:
    source, origins = apply_exceptions(prepare_text_mapped(text, options.utf16_offsets), options)
    output: List[str] = []
    warnings: List[Warning] = []
    previous_letter = ""
    position = 0

    for index, character in enumerate(source):
        lower = lower_char(character)
        replacement: Optional[str]

        if lower == "е":
            use_ye = (
                previous_letter == ""
                or previous_letter in _CYRILLIC_VOWELS
                or previous_letter == "ъ"
                or previous_letter == "ь"
            )
            replacement = "ye" if use_ye else "e"
            warnings.append(
                _warning(
                    position,
                    "cyrillic.e.positional",
                    "Cyrillic е can represent e or ye; a positional rule was applied.",
                    ("e", "ye"),
                )
            )
        elif lower == "ц":
            use_s = previous_letter == "" or previous_letter == "ц"
            replacement = "s" if use_s else "ts"
            warnings.append(
                _warning(
                    position,
                    "cyrillic.tse.positional",
                    "Cyrillic ц can represent s or ts; a positional rule was applied.",
                    ("s", "ts"),
                )
            )
        elif lower == "щ":
            replacement = "şç"
            warnings.append(
                _warning(
                    position,
                    "cyrillic.shcha.ambiguous",
                    "Cyrillic щ has no single new-Latin equivalent; şç was chosen.",
                    ("şç", "ş"),
                )
            )
        elif lower == "ъ":
            replacement = "ʼ"
            warnings.append(
                _warning(
                    position,
                    "cyrillic.hard-sign.ambiguous",
                    "The hard sign was retained as tutuq belgisi.",
                    ("ʼ", ""),
                )
            )
        elif lower == "ь":
            # Before a plain vowel the soft sign marks a glide (батальон -> batalyon); elsewhere
            # it has no Latin counterpart and is dropped (сентябрь -> medal).
            glide = _letter_at(source, index + 1) in _SOFT_SIGN_GLIDE_VOWELS
            replacement = "y" if glide else ""
            warnings.append(
                _warning(
                    position,
                    "cyrillic.soft-sign.ambiguous",
                    "The soft sign before a vowel was written as y."
                    if glide
                    else "The soft sign was dropped.",
                    ("y", "") if glide else ("", "ʼ"),
                )
            )
        elif lower in _IOTATED:
            name = _IOTATED[lower]
            replacement = name
            warnings.append(
                _warning(
                    position,
                    f"cyrillic.{name}.compound",
                    f"Cyrillic {lower} was expanded to {name}.",
                    (name, name[1:]),
                )
            )
        else:
            replacement = _DIRECT_FROM_CYRILLIC.get(lower)

        output.append(
            character if replacement is None else _case_replacement(source, index, replacement)
        )
        previous_letter = _letter_at(source, index)
        position += unit_length(character, options.utf16_offsets)

    warnings = _to_source_offsets(warnings, origins, offset)
    return nfc(old_latin_core("".join(output), options)), warnings


def _from_cyrillic(text: str, options: Options) -> ConversionResult:
    converted, warnings = map_segments(
        text, options, lambda segment, start: _from_cyrillic_core(segment, start, options)
    )
    return ConversionResult(converted, tuple(warnings))


def _to_cyrillic_core(text: str, offset: int, options: Options) -> Tuple[str, List[Warning]]:
    source, origins = prepare_text_mapped(text, options.utf16_offsets)
    length = len(source)
    output: List[str] = []
    warnings: List[Warning] = []
    index = 0
    position = 0
    while index < length:
        character = source[index]
        lower = lower_char(character)
        uppercase = character != lower
        pair = lower + (lower_char(source[index + 1]) if index + 1 < length else "")
        # Inverse of the fromCyrillic е rule: Cyrillic е is read "ye" at a word start, after
        # a vowel, or after ъ/ь, so there "ye" maps back to е and a bare "e" must be э.
        previous = _letter_at(source, index - 1) if index > 0 else ""
        positional = previous == "" or previous == "ʼ" or previous in _LATIN_VOWELS
        replacement: Optional[str]
        consumed = 1

        if pair == "şç":
            replacement = "щ"
            consumed = 2
            warnings.append(
                _warning(
                    position,
                    "latin.shcha.ambiguous",
                    "şç was interpreted as Cyrillic щ rather than шч.",
                    ("щ", "шч"),
                    2,
                )
            )
        elif pair in _IOTATED_REVERSE:
            replacement = _IOTATED_REVERSE[pair]
            consumed = 2
            warnings.append(
                _warning(
                    position,
                    f"latin.{pair}.ambiguous",
                    f"{pair} was interpreted as one iotated Cyrillic letter.",
                    (replacement, "й" + _DIRECT_TO_CYRILLIC[pair[1]]),
                    2,
                )
            )
        elif pair == "ts":
            replacement = "ц"
            consumed = 2
            warnings.append(
                _warning(
                    position,
                    "latin.tse.ambiguous",
                    "ts was interpreted as Cyrillic ц.",
                    ("ц", "тс"),
                    2,
                )
            )
        elif pair == "ye" and positional:
            replacement = "е"
            consumed = 2
            warnings.append(
                _warning(
                    position,
                    "latin.ye.positional",
                    "ye at a word start, after a vowel, or after tutuq was interpreted as "
                    "Cyrillic е.",
                    ("е", "йе"),
                    2,
                )
            )
        elif lower == "e":
            replacement = "э" if positional else "е"
            warnings.append(
                _warning(
                    position,
                    "latin.e.ambiguous",
                    f"Latin e can correspond to Cyrillic е or э; {replacement} was chosen.",
                    ("э", "е") if positional else ("е", "э"),
                )
            )
        elif character == "ʼ":
            replacement = "ъ"
            warnings.append(
                _warning(
                    position,
                    "latin.tutuq.ambiguous",
                    "Tutuq belgisi was interpreted as a hard sign.",
                    ("ъ", "ь", ""),
                )
            )
        elif lower == "c":
            replacement = "ц"
            warnings.append(
                _warning(
                    position,
                    "latin.c.ambiguous",
                    "Standalone c was interpreted as Cyrillic ц.",
                    ("ц", "с"),
                )
            )
        else:
            replacement = _DIRECT_TO_CYRILLIC.get(lower)
            if replacement is None and (_is_latin_letter(character) or character == "ʻ"):
                warnings.append(
                    _warning(
                        position,
                        "latin.unmapped",
                        (
                            "Stray ʻ is not part of oʻ or gʻ and has no Cyrillic mapping."
                            if character == "ʻ"
                            else f"Latin {character} has no Cyrillic mapping."
                        ),
                        (),
                    )
                )

        if replacement is None:
            output.append(character)
        elif uppercase:
            output.append(first_case(character, replacement))
        else:
            output.append(replacement)
        # A two-character match only ever consumes Basic Multilingual Plane letters.
        position += consumed if consumed == 2 else unit_length(character, options.utf16_offsets)
        index += consumed
    return nfc("".join(output)), _to_source_offsets(warnings, origins, offset)


def _to_cyrillic(text: str, options: Options) -> ConversionResult:
    canonical = to_new_latin_mapped(text, options)
    converted, warnings = map_segments(
        canonical.text,
        options,
        lambda segment, start: _to_cyrillic_core(segment, start, options),
    )
    # Warnings point into the new-Latin intermediate; report them against the caller's text.
    return ConversionResult(converted, tuple(_to_source_offsets(warnings, canonical.origins, 0)))


def from_cyrillic(
    text: str,
    *,
    protect_spans: bool = True,
    protected_terms: Optional[Iterable[str]] = None,
    exceptions: Optional[Mapping[str, str]] = None,
    ng_as_digraph: bool = True,
) -> ConversionResult:
    """Convert Uzbek Cyrillic to new Latin and report every lossy or ambiguous choice."""
    return _from_cyrillic(
        text,
        make_options(
            protect_spans=protect_spans,
            protected_terms=protected_terms,
            exceptions=exceptions,
            ng_as_digraph=ng_as_digraph,
        ),
    )


def to_cyrillic(
    text: str,
    *,
    protect_spans: bool = True,
    protected_terms: Optional[Iterable[str]] = None,
    exceptions: Optional[Mapping[str, str]] = None,
    ng_as_digraph: bool = True,
) -> ConversionResult:
    """Convert new Uzbek Latin to Cyrillic and report every lossy or ambiguous choice."""
    return _to_cyrillic(
        text,
        make_options(
            protect_spans=protect_spans,
            protected_terms=protected_terms,
            exceptions=exceptions,
            ng_as_digraph=ng_as_digraph,
        ),
    )
